File: code/R/do_analysis.R
# ------------------------------------------------------------------------------
# Prepares figures and tables for our showcase project, exploring discretionary
# accruals.
#
# See LICENSE file for licensing information.
# ------------------------------------------------------------------------------


# --- Read config --------------------------------------------------------------

source("code/R/utils.R")


# --- Read TRR 266 ggplot2 theme -----------------------------------------------

source("code/R/theme_trr.R")


# --- Create tables and figures ------------------------------------------------

log_info("Reading and preparing data...")
smp <- read_parquet(global_cfg$acc_sample)

graph_boxplot <- function(df) {
  df <- df %>%
    select(fyear, mj_da, dd_da) %>%
    filter(!is.na(dd_da)) %>%
    pivot_longer(
      any_of(c("mj_da", "dd_da")), names_to = "type", values_to = "da"
    )

  ggplot(
    df,
    aes(x = fyear, y = da, group = interaction(type, fyear), color = type)
  ) +
    geom_boxplot() +
    labs(
      x = "Fiscal year", y = NULL, color = "Type of discretionary accruals"
    ) +
    scale_color_trr266_d(labels = c("Dechow and Dichev", "Modified Jones")) +
    theme_trr(legend = TRUE)
}

smp_da <- smp %>%
  select(
    gvkey, fyear, ff12_ind, mj_da, dd_da, ln_ta, ln_mktcap, mtb,
    ebit_avgta, sales_growth
  )

smp_da <- smp_da[is.finite(rowSums(smp_da %>% select(-gvkey, -ff12_ind))), ]

smp_da <- treat_outliers(smp_da, by = "fyear")

log_info("Preparing figures...")

fig_boxplot_full <- graph_boxplot(smp)
fig_boxplot_smp <- graph_boxplot(smp_da)

fig_scatter_md_dd <- ggplot(smp_da, aes(x = mj_da, y = dd_da)) +
  geom_bin2d(
    aes(fill = stat(log(count))),  bins = c(100, 100)
  ) +
  labs(x = "Modified Jones DA", y = "Dechow and Dichev DA") +
  scale_fill_trr266_c() +
  theme_trr(axis_y_horizontal = FALSE)

fig_scatter_dd_lnta <- ggplot(smp_da, aes(x = ln_ta, y = dd_da)) +
  geom_bin2d(
    aes(fill = stat(log(count))),  bins = c(100, 100)
  ) +
  labs(x = "ln(Total Assets)", y = "Dechow and Dichev DA") +
  scale_fill_trr266_c() +
  theme_trr(axis_y_horizontal = FALSE)

fig_scatter_dd_roa <- ggplot(smp_da, aes(x = ebit_avgta, y = dd_da)) +
  geom_bin2d(
    aes(fill = stat(log(count))),  bins = c(100, 100)
  ) +
  labs(x = "Return on Assets", y = "Dechow and Dichev DA") +
  scale_fill_trr266_c() +
  theme_trr(axis_y_horizontal = FALSE)

fig_scatter_dd_salesgr <- ggplot(smp_da, aes(x = sales_growth, y = dd_da)) +
  geom_bin2d(
    aes(fill = stat(log(count))),  bins = c(100, 100)
  ) +
  labs(x = "Sales Growth", y = "Dechow and Dichev DA") +
  scale_fill_trr266_c() +
  theme_trr(axis_y_horizontal = FALSE)

log_info("Preparing tables...")

var_names <- tibble(
  var_name = names(smp_da %>% select(-gvkey, -fyear, -ff12_ind)),
  label = cdesc_rnames <- c(
    "Modified Jones DA",
    "Dechow and Dichev DA",
    "Ln(Total assets)",
    "Ln(Market capitalization)",
    "Market to book",
    "Return on assets",
    "Sales growth"
  )
)

N <- function(x) (format(sum(!is.na(x)), big.mark = ","))
desc_smp <- smp_da %>% select(-c(fyear, gvkey, ff12_ind))
colnames(desc_smp) <- var_names$label
tab_desc_stat <- datasummary(
  All(desc_smp) ~ N + Mean + SD + Min + P25 + Median + P75 + Max, 
  data = desc_smp, fmt = 3, output = "gt"
)


rlabels <- paste0(
  LETTERS[1:nrow(var_names)], ": ", colnames(desc_smp)
)
clabels <- c(" ", LETTERS[1:nrow(var_names)])
names(clabels) <- c(" ", rlabels)
colnames(desc_smp) <- rlabels
tab_corr <-datasummary_correlation(
  desc_smp, method = "pearspear", output = "gt"
) %>% cols_label(!!!clabels)

mods <- feols(
  c(mj_da, dd_da) ~ ln_ta + mtb + ebit_avgta + sales_growth | gvkey + fyear,
  smp_da, cluster = c("gvkey", "fyear")
)
for (i in 1:2) {
  mods[[i]]$singletons <- mods[[i]]$nobs_orig - mods[[i]]$nobs
}

names(mods) <- var_names$label[1:2]
ar <- tibble(
  col1 = c(
    "Observations", "Singletons dropped", "Observations used", 
    "Fixed Effects", "SE Clustered"
  ),
  col2 = c(
    format(mods[[1]]$nobs_orig, big.mark = ","),
    format(mods[[1]]$singletons, big.mark = ","),
    format(mods[[1]]$nobs, big.mark = ","),
    "Firm and Year", "Firm and Year"
  ),
  col3 = c(
    format(mods[[2]]$nobs_orig, big.mark = ","),
    format(mods[[2]]$singletons, big.mark = ","),
    format(mods[[2]]$nobs, big.mark = ","),
    "Firm and Year", "Firm and Year"
  )
)
attr(ar, "position") <- 9:13

tab_regression <- modelsummary(
  mods, stars = c(`***` = 0.01, `**` = 0.05, `*` = 0.1),
  estimate = "{estimate}{stars}",
  add_rows = ar,
  gof_map = list(
    list(
      raw = "adj.r.squared", 
      clean = "Adj. R² (overall)", 
      fmt = function(x) sprintf("%.3f", x)
    ),
    list(
      raw = "r2.within.adjusted", clean = "Adj. R² (within)", 
      fmt = function(x) sprintf("%.3f", x)
    )
  ),
  coef_rename = var_names$label[c(3, 5:7)],
  output = "gt"
)

log_info("Done. Storing output objects in '{global_cfg$results_r}'")
save(
  list = c(
    "smp_da", "var_names", ls(pattern = "^fig_*"), ls(pattern = "^tab_*")
  ),
  file = global_cfg$results_r
)

File: code/python/do_analysis.py
import pickle
import pandas as pd
import numpy as np
from plotnine import ggplot, geom_boxplot, labs, aes, geom_bin2d, after_stat
from panel_eda_helper_funcs import (prepare_correlation_table,
                                    prepare_descriptive_table, PrepareRegressionTable,
                                    escape_for_latex)
from prepare_data import treat_outliers
from theme_trr import theme_trr, scale_color_trr266_d, scale_fill_trr266_c

from utils import setup_logging, read_config

log = setup_logging()
global_cfg = read_config('config/global_cfg.yaml')

def main():
    log.info("Reading and preparing data...")
    smp = pd.read_parquet(global_cfg['acc_sample'])
    smp_da = prep_smp_da(smp)

    log.info("Preparing figures...")
    fig_boxplot_full = make_boxplot(smp)


    fig_boxplot_smp = make_boxplot(smp_da)

    fig_scatter_md_dd = make_scatter_plot(
        smp_da, 'mj_da', 'dd_da', 'Modified Jones DA', 'Dechow and Dichev DA'
    )

    fig_scatter_dd_lnta = make_scatter_plot(
        smp_da, 'ln_ta', 'dd_da', 'ln(Total Assets)', 'Dechow and Dichev DA'
    )

    fig_scatter_dd_roa = make_scatter_plot(
        smp_da, 'ebit_avgta', 'dd_da', 'Return on Assets', 'Dechow and Dichev DA'
    )
    fig_scatter_dd_salesgr = make_scatter_plot(
        smp_da, 'sales_growth', 'dd_da', 'Sales Growth', 'Dechow and Dichev DA'
    )


    log.info("Preparing tables...")
    tab_desc_stat = prepare_descriptive_table(smp_da.drop(['fyear'], axis=1))

    desc_info = {
        'min_fyear': smp_da['fyear'].min(),
        'max_fyear': smp_da['fyear'].max(),
        'no_unique_firms': smp_da['gvkey'].nunique()
    }

    tab_corr = prepare_correlation_table(smp_da.drop(['fyear'], axis=1))

    smp_da = smp_da.set_index(['gvkey', 'fyear'])

    tab_regression = PrepareRegressionTable(
        smp_da,
        dvs=['mj_da', 'dd_da'],
        idvs=[
            ['ln_ta', 'mtb', 'ebit_avgta', 'sales_growth'],
            ['ln_ta', 'mtb', 'ebit_avgta', 'sales_growth']
        ],
        entity_effects=[True, True], time_effects=[True, True],
        cluster_entity=[True, True], cluster_time=[True, True],
        models=['ols', 'ols']
    ).latex_table

    var_names = list(smp_da.drop('ff12_ind', axis=1).columns.values)
    label = [
        'Modified Jones DA',
        'Dechow and Dichev DA',
        'Ln(Total assets)',
        'Ln(Market capitalization)',
        'Market to book',
        'Return on assets',
        'Sales growth'
    ]

    var_names = dict(zip([escape_for_latex(var) for var in var_names], label))

    results = {
        'fig_boxplot_full': fig_boxplot_full,
        'fig_boxplot_smp': fig_boxplot_smp,
        'fig_scatter_md_dd': fig_scatter_md_dd,
        'fig_scatter_dd_lnta': fig_scatter_dd_lnta,
        'fig_scatter_dd_roa': fig_scatter_dd_roa,
        'fig_scatter_dd_salesgr': fig_scatter_dd_salesgr,
        'Descriptive Statistics': tab_desc_stat,
        'Desc information': desc_info,
        'Correlations': tab_corr,
        'Regression': tab_regression,
        'Variable Names': var_names
    }
    log.info(f"Done. Storing output objects in '{global_cfg['results_r']}'")
    with open(global_cfg['results_python'], 'wb') as f:
        pickle.dump(results, f)


def make_boxplot(df):
    df = df \
        .filter(['fyear', 'mj_da', 'dd_da']) \
        .query('dd_da.notna()') \
        .melt('fyear', ['mj_da', "dd_da"], 'type', 'da') \
        .assign(types=lambda x: x[['fyear', 'type']].astype(str)
                .agg('.'.join, axis=1))

    return (ggplot(df, aes(x='fyear', y='da', group='types', color='type')) +
            geom_boxplot() +
            labs(x="Fiscal year", y=None, color="Type of discretionary accruals") +
            scale_color_trr266_d(labels=["Dechow and Dichev", "Modified Jones"]) +
            theme_trr(legend=True)).draw()


def prep_smp_da(smp):
    smp_da = smp[['gvkey', 'fyear', 'ff12_ind', 'mj_da', 'dd_da',
                  'ln_ta', 'ln_mktcap', 'mtb', 'ebit_avgta', 'sales_growth']]

    smp_da = smp_da[
        np.isfinite(smp_da.drop(['gvkey', 'fyear'], axis=1).sum(
            numeric_only=True, axis=1, skipna=False))
    ]
    smp_da = treat_outliers(smp_da, by="fyear")
    return smp_da


def make_scatter_plot(df, x, y, xlab, ylab):
    return (ggplot(df, aes(x=x, y=y)) +
            geom_bin2d(aes(fill=after_stat('np.log(count)')), bins=[100, 100]) +
            labs(x=xlab, y=ylab) +
            scale_fill_trr266_c() +
            theme_trr(axis_y_horizontal=False)).draw()


if __name__ == "__main__":
    main()
